port/src/devicetree.rs

```diff
@@ -0,0 +1,738 @@
+use core::ffi::CStr;
+
+#[derive(Debug)]
+pub enum ParseError {
+    InvalidHeader,
+    InvalidMagic,
+    BufferTooSmall,
+    InvalidToken,
+}
+
+type Result<T> = core::result::Result<T, ParseError>;
+
+fn bytes_to_u32(bytes: &[u8]) -> Option<u32> {
+    Some(u32::from_be_bytes(bytes.get(..4)?.try_into().unwrap()))
+}
+
+fn bytes_to_u64(bytes: &[u8]) -> Option<u64> {
+    Some(u64::from_be_bytes(bytes.get(..8)?.try_into().unwrap()))
+}
+
+/// Extract u32 from bytes, but cast as u64
+fn bytes_to_u32_as_u64(bytes: &[u8]) -> Option<u64> {
+    Some(u32::from_be_bytes(bytes.get(..4)?.try_into().unwrap()) as u64)
+}
+
+fn align4(n: usize) -> usize {
+    let mut aligned_n = n;
```

Comment:
My personal favorite way to express this:

    n + (0_usize.wrapping_sub(n) & 3)